File: src/evaluation/generate_full_cohort_predictions.R
#!/usr/bin/env Rscript
# Generate Full Cohort RSF Predictions for MIMIC-IV Data
# =======================================================
# Trains RSF on full n=852 cohort and generates predictions for all patients

library(survival)
library(randomForestSRC)
library(dplyr)

cat("================================================================================\n")
cat("GENERATING FULL COHORT RSF PREDICTIONS (n=852)\n")
cat("================================================================================\n\n")

# ============================================================================
# 1. LOAD FULL MIMIC-IV COHORT
# ============================================================================
cat("1. Loading full MIMIC-IV cohort...\n")

data <- read.csv("mimic_sepsis_cohort_full.csv")
cat(sprintf("   Loaded: mimic_sepsis_cohort_full.csv\n"))
cat(sprintf("   Total patients: n=%d\n", nrow(data)))
cat(sprintf("   Total events: %d (%.1f%%)\n", sum(data$Event), 100 * mean(data$Event)))

# ============================================================================
# 2. PREPARE DATA FOR RSF
# ============================================================================
cat("\n2. Preparing data for RSF...\n")

# Remove ID columns
feature_cols <- setdiff(names(data), c("subject_id", "hadm_id", "Time", "Event"))
cat(sprintf("   Features: %d variables\n", length(feature_cols)))
cat(sprintf("   Feature names: %s...\n", paste(feature_cols[1:min(5, length(feature_cols))], collapse = ", ")))

# Handle missing values (impute with median for numeric, mode for categorical)
for (col in feature_cols) {
    if (is.numeric(data[[col]])) {
        data[[col]][is.na(data[[col]])] <- median(data[[col]], na.rm = TRUE)
    } else {
        mode_val <- names(sort(table(data[[col]]), decreasing = TRUE))[1]
        data[[col]][is.na(data[[col]])] <- mode_val
    }
}

# Create formula for RSF
formula_str <- paste("Surv(Time, Event) ~", paste(feature_cols, collapse = " + "))
cat(sprintf("   Formula: %s\n", substr(formula_str, 1, 60)))

# ============================================================================
# 3. TRAIN RSF ON FULL COHORT
# ============================================================================
cat("\n3. Training RSF on full cohort...\n")
cat("   (This may take a few minutes...)\n")

rsf_model <- rfsrc(
    as.formula(formula_str),
    data = data,
    ntree = 1000,
    nodesize = 15,
    splitrule = "logrank",
    importance = TRUE,
    seed = 42
)

cat("   RSF training complete!\n")
cat(sprintf("   OOB C-index: %.3f\n", rsf_model$err.rate[length(rsf_model$err.rate)]))

# ============================================================================
# 4. GENERATE PREDICTIONS FOR ALL PATIENTS
# ============================================================================
cat("\n4. Generating predictions for all patients...\n")

# Get predicted mortality (higher = higher risk)
risk_scores <- rsf_model$predicted.oob

# Create risk groups (median split)
median_risk <- median(risk_scores)
risk_groups <- ifelse(risk_scores >= median_risk, "High", "Low")

# Prepare output dataframe
predictions <- data.frame(
    Patient_ID = 1:nrow(data),
    Observed_Time = data$Time,
    Observed_Status = data$Event,
    Predicted_Risk = risk_scores,
    Risk_Group = risk_groups
)

# Save predictions
write.csv(predictions, "full_cohort_rsf_predictions.csv", row.names = FALSE)
cat(sprintf("   Saved: full_cohort_rsf_predictions.csv (n=%d)\n", nrow(predictions)))

# ============================================================================
# 5. SUMMARY STATISTICS
# ============================================================================
cat("\n5. Summary statistics:\n")

for (group in c("High", "Low")) {
    subset <- predictions[predictions$Risk_Group == group, ]
    cat(sprintf("\n   %s Risk Group:\n", group))
    cat(sprintf("     N = %d\n", nrow(subset)))
    cat(sprintf("     Events = %d\n", sum(subset$Observed_Status)))
    cat(sprintf("     Event Rate = %.1f%%\n", 100 * mean(subset$Observed_Status)))
    cat(sprintf("     Mean Risk Score = %.2f\n", mean(subset$Predicted_Risk)))
}

cat("\n" + "=" * 80)
cat("PREDICTION GENERATION COMPLETE\n")
cat("================================================================================\n")
cat("\nNext step:\n")
cat("  The predictions are ready in: full_cohort_rsf_predictions.csv\n")
cat("  Now calculating survival statistics...\n")
cat("================================================================================\n\n")

# ============================================================================
# 6. CALCULATE SURVIVAL STATISTICS IMMEDIATELY
# ============================================================================
cat("6. Calculating survival statistics on full cohort...\n\n")

# Calibration
predictions$risk_decile <- cut(
    predictions$Predicted_Risk,
    breaks = quantile(predictions$Predicted_Risk, probs = seq(0, 1, 0.1)),
    labels = FALSE,
    include.lowest = TRUE
)

calib <- predictions %>%
    group_by(risk_decile) %>%
    summarise(
        Mean_Predicted_Risk = mean(Predicted_Risk),
        Observed_Event_Rate = mean(Observed_Status)
    )

write.csv(calib, "full_cohort_calibration.csv", row.names = FALSE)
cat(sprintf(
    "   Calibration Range: %.2f%% - %.2f%%\n",
    min(calib$Mean_Predicted_Risk), max(calib$Mean_Predicted_Risk)
))

# Survival Probabilities
results_surv <- data.frame()

for (group in c("High", "Low")) {
    sub <- predictions[predictions$Risk_Group == group, ]
    fit <- survfit(Surv(Observed_Time, Observed_Status) ~ 1, data = sub)
    summ <- summary(fit, times = c(30, 60), extend = TRUE)

    row <- data.frame(
        Group = group,
        N = nrow(sub),
        Events = sum(sub$Observed_Status),
        Surv_30d = if (length(summ$surv) >= 1) summ$surv[1] else NA,
        CI_30d_Lower = if (length(summ$lower) >= 1) summ$lower[1] else NA,
        CI_30d_Upper = if (length(summ$upper) >= 1) summ$upper[1] else NA,
        Surv_60d = if (length(summ$surv) >= 2) summ$surv[2] else NA,
        CI_60d_Lower = if (length(summ$lower) >= 2) summ$lower[2] else NA,
        CI_60d_Upper = if (length(summ$upper) >= 2) summ$upper[2] else NA
    )
    results_surv <- rbind(results_surv, row)
}

write.csv(results_surv, "full_cohort_survival_probs.csv", row.names = FALSE)
cat("\n   Survival Probabilities:\n")
print(results_surv)

# Hazard Ratio
predictions$is_high_risk <- ifelse(predictions$Risk_Group == "High", 1, 0)
cox_fit <- coxph(Surv(Observed_Time, Observed_Status) ~ is_high_risk, data = predictions)
hr <- exp(coef(cox_fit))
ci <- exp(confint(cox_fit))
p <- summary(cox_fit)$coefficients[, "Pr(>|z|)"]

hr_result <- data.frame(
    HR = hr,
    CI_Lower = ci[1],
    CI_Upper = ci[2],
    P_Value = p,
    Significant = ifelse(p < 0.05, "YES", "NO")
)

write.csv(hr_result, "full_cohort_hazard_ratio.csv", row.names = FALSE)
cat(sprintf("\n   HR: %.2f (95%% CI: %.2f-%.2f), p=%.4f\n", hr, ci[1], ci[2], p))
cat(sprintf("   Statistically significant: %s\n\n", hr_result$Significant))

cat("================================================================================\n")
cat("ALL DONE! Results saved to CSV files.\n")
cat("================================================================================\n")

File: src/evaluation/generate_full_cohort_predictions.py
"""
Generate Full Cohort Predictions for Survival Analysis
=======================================================
This script trains RSF on the FULL cohort (n=852) and generates
predictions for ALL patients to enable proper survival analysis.

Input: synthetic_complete.csv (or your real MIMIC-IV data)
Output: full_cohort_rsf_predictions.csv with n=852 predictions
"""

import pandas as pd
import numpy as np
import os
from sklearn.preprocessing import StandardScaler
from sksurv.util import Surv
from sksurv.ensemble import RandomSurvivalForest

print("=" * 80)
print("GENERATING FULL COHORT RSF PREDICTIONS (n=852)")
print("=" * 80)

# ============================================================================
# 1. LOAD FULL COHORT DATA
# ============================================================================
print("\n1. Loading full cohort data...")

# Updated paths to check for the data file
possible_data_paths = [
    "mimic_sepsis_cohort_full.csv",
    "data/raw/mimic_sepsis_cohort_full.csv",
    "../data/raw/mimic_sepsis_cohort_full.csv",
    "../../data/raw/mimic_sepsis_cohort_full.csv"
]

data = None
for path in possible_data_paths:
    if os.path.exists(path):
        try:
            data = pd.read_csv(path)
            print(f"   Loaded: {path}")
            break
        except Exception as e:
            print(f"   Error reading {path}: {e}")

if data is None:
    print("   ERROR: mimic_sepsis_cohort_full.csv not found!")
    print("   Please ensure the data file exists in one of the following locations:")
    for path in possible_data_paths:
        print(f"     - {path}")
    exit(1)

print(f"   Total patients: n={len(data)}")
print(f"   Total events: {data['Event'].sum()} ({100*data['Event'].mean():.1f}%)")

# ============================================================================
# 2. PREPARE DATA
# ============================================================================
print("\n2. Preparing data for RSF...")

# Separate features and outcome
feature_cols = [col for col in data.columns if col not in ['Time', 'Event']]
X = data[feature_cols].values
y = Surv.from_dataframe('Event', 'Time', data)

print(f"   Features: {len(feature_cols)} variables")
print(f"   Feature names: {', '.join(feature_cols[:5])}...")

# Standardize features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# ============================================================================
# 3. TRAIN RSF ON FULL COHORT
# ============================================================================
print("\n3. Training RSF on full cohort...")

# CRITICAL FIX: Add oob_score=True to generate unbiased predictions for training data
rsf = RandomSurvivalForest(
    n_estimators=1000,
    min_samples_split=10,
    min_samples_leaf=15,
    n_jobs=-1,
    random_state=42,
    oob_score=True   # Added to enable OOB predictions
)

rsf.fit(X_scaled, y)
print("   RSF training complete!")

# ============================================================================
# 4. GENERATE PREDICTIONS FOR ALL PATIENTS
# ============================================================================
print("\n4. Generating predictions for all patients...")

# CRITICAL FIX: Use unbiased OOB predictions instead of biased rsf.predict(X_scaled)
try:
    risk_scores = rsf.oob_prediction_
    print("   Uising UNBIASED Out-of-Bag (OOB) risk scores.")
except AttributeError:
    print("   WARNING: OOB predictions not found. Falling back to standard prediction (LEAKAGE RISK).")
    risk_scores = rsf.predict(X_scaled)

# Create risk groups (median split)
median_risk = np.median(risk_scores)

# Use >= for High to handle potential ties; check split counts
risk_groups = ["High" if r >= median_risk else "Low" for r in risk_scores]

# Prepare output dataframe
predictions = pd.DataFrame({
    'Patient_ID': range(len(data)),
    'Observed_Time': data['Time'].values,
    'Observed_Status': data['Event'].values.astype(int),
    'Predicted_Risk': risk_scores,
    'Risk_Group': risk_groups
})

# Save predictions
predictions.to_csv("full_cohort_rsf_predictions.csv", index=False)
print(f"   Saved: full_cohort_rsf_predictions.csv (n={len(predictions)})")

# ============================================================================
# 5. SUMMARY STATISTICS
# ============================================================================
print("\n5. Summary statistics:")

for group in ['High', 'Low']:
    subset = predictions[predictions['Risk_Group'] == group]
    print(f"\n   {group} Risk Group:")
    print(f"     N = {len(subset)}")
    print(f"     Events = {subset['Observed_Status'].sum()}")
    print(f"     Event Rate = {100*subset['Observed_Status'].mean():.1f}%")
    print(f"     Mean Risk Score = {subset['Predicted_Risk'].mean():.2f}")

print("\n" + "=" * 80)
print("PREDICTION GENERATION COMPLETE")
print("=" * 80)
print("\nNext step:")
print("  Run: python src/visualization/generate_km_curves.py")
print("=" * 80)